packages/dev/core/src/Materials/greasedLineBaseMaterial.ts

```diff
@@ -0,0 +1,348 @@
+     * Defaults to 0.1 if @see sizeAttenuation is false, or to 1 if it's true.
+     */
+    width?: number;
+    /**
+     * If false then width units = scene units. If true then line will width be reduced.
+     * Defaults to false.
+     */
+    sizeAttenuation?: boolean;
+    /**
+     * Type of the material to use to render the line.
+     * Defaults to StandardMaterial.
+     */
+    materialType?: GreasedLineMeshMaterialType;
+    /**
+     * Color of the line. Applies to all line segments.
+     * Defaults to White.
+     */
+    color?: Color3;
+    /**
+     * Color mode of the line. Applies to all line segments.
+     * The pixel color from the material shader will be modified with the value of @see color using the colorMode.
+     * Defaults to @see GreasedLineMeshColorMode.SET
+     */
+    colorMode?: GreasedLineMeshColorMode;
+    /**
+     * Colors of the line segments.
+     * Defaults to empty.
+     */
+    colors?: Color3[];
+    /**
+     * If true, @see colors are used, otherwise they're ignored.
+     * Defaults to false.
+     */
+    useColors?: boolean;
+    /**
+     * Sampling type of the colors texture
+     * Defaults to NEAREST_NEAREST.
+     */
+    colorsSampling?: number;
+    /**
+     * The method used to distribute the colors along the line.
+     * You can use segment distribution when each segment will use on color from the color table.
+     * Or you can use line distribution when the colors are distributed evenly along the line ignoring the segments.
+     */
+    colorDistributionType?: GreasedLineMeshColorDistributionType;
+    /**
+     * If true, dashing is used.
+     * Defaults to false.
+     */
+    useDash?: boolean;
+    /**
+     * @see GreasedLinePluginMaterial.setDashCount
+     * Number of dashes in the line.
+     * Defaults to 1.
+     */
+    dashCount?: number;
+    /**
+     * Offset of the dashes along the line. 0 to 1.
+     * Defaults to 0.
+     * @see GreasedLinePluginMaterial.setDashOffset
+     */
+    dashOffset?: number;
+    /**
+     * Length of the dash. 0 to 1. 0.5 means half empty, half drawn.
+     * Defaults to 0.5.
+     * @see GreasedLinePluginMaterial.setDashRatio
+     */
+    dashRatio?: number;
+    /**
+     * Sets the line length visibility.
+     * 0 - 0% of the line will be visible.
+     * 1 - 100% of the line will be visible.
+     * @see GreasedLinePluginMaterial.setVisibility
+     */
+    visibility?: number;
+    /**
+     * Defaults to engine.getRenderWidth() and engine.getRenderHeight()
+     * Rendering resolution
+     */
+    resolution?: Vector2;
+    /**
+     * Whether to use camera facing for the line.
+     * Defaults to true.
+     */
+    cameraFacing?: boolean;
+    /**
+     * You can provide a colorsTexture to use instead of one generated from the 'colors' option
+     */
+    colorsTexture?: RawTexture;
+}
+
+/**
+ * Base class for GreasedLine materials
+ */
+export class GreasedLineBaseMaterial {

```

Comment:
Suggestion: Why not moving everything from this class to `GreasedLineTools` instead, as everything is static in `GreasedLineBaseMaterial` (in fact, `GreasedLineBaseMaterial` is not a base class for the greased line material calsses)? If you do it, `greasedLineBaseMaterial.ts` can be renamed to something like `greasedLineInterfaces.ts`, as only interfaces/type declarations will remain in the file.

Reply:
Adding it to `GreasedLineBuilder` would be ok I think.

However, we now have 3 files in the `Materials/` folder, and 4 in the `Meshes/` folder (counting the builder file) for the new feature. I wonder if we shouldn't create a sub-directory `Materials/GreasedLine/` and `Meshes/GreasedLine/` to make things clearer?

What do you think @sebavan?

Reply:
agree, could be good to group them.

Reply:
@Popov72 Now that we have a dedicated folder for greased line I left the `GreasedLineMaterialDefaults` class in a standalone file.